File: factory_pattern_py/test_shape_factory.py
import unittest 
from ShapeFactory import ShapeFactory

class ShapeFactoryRectangleTestCase(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.shapeFactory = ShapeFactory()
        
    def setUp(self):
        self.myShape = self.shapeFactory.getShape('RECTANGLE')

    def test_get_area_dim1(self):
        self.myShape.setDim([10,20])
        self.assertEqual(2,self.myShape.d_size())
        self.assertEqual(200,self.myShape.getArea())

class ShapeFactorySquareTestCase(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.shapeFactory = ShapeFactory()

    def setUp(self):
        self.myShape = self.shapeFactory.getShape('SQUARE')

    def test_get_area_dim1(self):
        self.myShape.setDim([20])
        self.assertEqual(1,self.myShape.d_size())
        self.assertEqual(400,self.myShape.getArea())

class ShapeFactoryCircleTestCase(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.shapeFactory = ShapeFactory()
    def setUp(self):
        self.myShape = self.shapeFactory.getShape('CIRCLE')

    def test_get_area_dim1(self):
        self.myShape.setDim([2])
        self.assertEqual(1,self.myShape.d_size())
        self.assertEqual(12.56,self.myShape.getArea())

    def test_get_area_dim2(self):
        self.myShape.setDim([1])
        self.assertEqual(1,self.myShape.d_size())
        self.assertEqual(3.14,self.myShape.getArea())

File: factory_pattern_cpp/test_shape_factory.cc
#include <gtest/gtest.h>
#include "ShapeFactory.h"
#include "Shape.h"

class ShapeFactoryTestCase: public ::testing::Test { 
  protected: 
   static void SetUpTestCase( ) { 
  
   } 
   static void TearDownTestCase( ) { 
      
   } 
    ShapeFactory shapeFactory;
};

class ShapeFactoryRectangleTestCase: public ShapeFactoryTestCase { 
protected: 
   void SetUp() override { 
     myShape = shapeFactory.getShape(ShapeFactory::ShapeType::Rectangle);
   }
   void TearDown() override { 
       delete myShape;
   }
    Shape * myShape;
};

class ShapeFactorySquareTestCase: public ShapeFactoryTestCase { 
protected: 
   void SetUp() override { 
     myShape = shapeFactory.getShape(ShapeFactory::ShapeType::Square);
   }
   void TearDown() override { 
       delete myShape;
   }
    Shape * myShape;
};

class ShapeFactoryCircleTestCase: public ShapeFactoryTestCase { 
protected: 
   void SetUp() override { 
     myShape = shapeFactory.getShape(ShapeFactory::ShapeType::Circle);
   }
   void TearDown() override { 
       delete myShape;
   }
    Shape * myShape;
};


TEST_F(ShapeFactoryRectangleTestCase, TEST_GET_AREA_DIM1) {
  double d[] = {10,20};
  myShape->setDim(d);
  EXPECT_EQ(2, myShape->d_size());
  EXPECT_EQ(200, myShape->getArea());
}

TEST_F(ShapeFactorySquareTestCase, TEST_GET_AREA_DIM1) { 
  double d[] = {20};
  myShape->setDim(d);
  EXPECT_EQ(1, myShape->d_size());
  EXPECT_EQ(400, myShape->getArea());
}

TEST_F(ShapeFactoryCircleTestCase, TEST_GET_AREA_DIM1) { 
  double d[] = {2};
  myShape->setDim(d);
  EXPECT_EQ(1, myShape->d_size());
  EXPECT_EQ(12.56, myShape->getArea());
}

TEST_F(ShapeFactoryCircleTestCase, TEST_GET_AREA_DIM2) { 
  double d[] = {1};
  myShape->setDim(d);
  EXPECT_EQ(1, myShape->d_size());
  EXPECT_EQ(3.14, myShape->getArea());
}